### PR TITLE
use watchman to orchestrate the development lifecycle

### PR DESCRIPTION
**Pull request by wenzowski**
...there should probably be a note in this that you need to actually install [watchman](https://facebook.github.io/watchman/) for this to work...

### CODE REVIEW

**Comment by matthieu-foucault** [comment]
Add info about this in the [developer guide](https://github.com/bcgov/cas-ciip-portal/blob/develop/documentation/developer-guide.md) ? :smile:

**Comment by wenzowski** [comment]
I think this is currently broken because  `sqitch rebase -y` is failing to complete. Have to dash but there should be answers in `logs/sqitch.stderr.log`

File: Makefile [hunk 1]
```diff
@@ -69,3 +69,28 @@ install:
...
+unwatch:
+	watchman watch-del-all
+	watchman shutdown-server
+	if command -v launchctl; then launchctl unload ~/Library/LaunchAgents/com.github.facebook.watchman.plist; fi
+
+# TODO: delete this target
+# ...but maybe it's helpful for debugging watchman itself on a mac?
+.PHONY: watch_log
+watch_log:
+	tail -f /usr/local/var/run/watchman/$(shell whoami)-state/log
```
Comment (matthieu-foucault) [inline Makefile]: `tail: cannot open '/usr/local/var/run/watchman/matthieu-state/log' for reading: No such file or directory`

Comment (wenzowski) [reply Makefile]: yes, that's the internal log for watchman itself

Comment (wenzowski) [reply Makefile]: we shouldn't be writing anything to it...unless we forgot to redirect our streams

Comment (wenzowski) [reply Makefile]: ideally each watcher is writing to `log/*` in the project dir

**Comment by wenzowski** [comment]
rebased on develop

File: app/server/schema.graphql [hunk 2]
```diff
@@ -1,15 +1,18 @@
 """The application data"""
```
Comment (matthieu-foucault) [inline app/server/schema.graphql]: This file probably shouldn't be updated by this PR

Comment (wenzowski) [reply app/server/schema.graphql]: I've changed the schema generation to be in alphabetic order. Could split if needed

Comment (wenzowski) [reply app/server/schema.graphql]: [`--sort-export`](https://www.graphile.org/postgraphile/usage-cli/#cli-options)
lexicographically (alphabetically) sort exported schema for more stable diffing.

File: app/package.json [hunk 3]
```diff
@@ -3,14 +3,15 @@
...
-    "pretest": "npm run relay",
+    "pretest": "npm run build",
     "test": "jest",
-    "predev": "npm run relay",
+    "predev": "npm run build",
     "dev": "node server",
-    "prebuild": "npm run relay",
-    "build": "next build",
+    "build": "npm run build:schema && npm run build:relay && npm run build:next",
+    "build:schema": "postgraphile -X -c postgres://localhost/ggircs_dev -s ggircs_portal --export-schema-graphql server/schema.graphql --export-schema-json server/schema.json --classic-ids --sort-export",
```
Comment (wenzowski) [inline app/package.json]: @matthieu-foucault `--sort-export` added here

Comment (wenzowski) [reply app/package.json]: ...though I probably shouldn't be hardcoding `postgres://localhost/ggircs_dev`...

Comment (matthieu-foucault) [reply app/package.json]: Running `build:schema` in the `build` script won't work on CI, as postgres is not running in the builder image

**Comment by matthieu-foucault** [comment]
@wenzowski `build_app` is a required check because it passes on `develop`, so it does need to pass for this PR to be mergeable :smiley:

File: app/package.json (hunk 3 above)
Comment (matthieu-foucault) [reply app/package.json]: I see that 02429bb79c4fee264cf3319fe00c11324b27bfd6 fixes this. Can you cherry pick it to this PR?

Comment (wenzowski) [reply app/package.json]: cherry-pick'd it over

File: .watch-app-format.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  "./app",
+  {
+    "name": "format",
+    "expression": [
+      "anyof",
+      ["match", "**/*.js", "wholename"],
+      ["match", "**/*.json", "wholename"],
+      ["match", "yarn.lock", "wholename"]
+    ],
+    "command": ["/usr/bin/env", "bash", "-c", "yarn run format 1>../log/format.stdout.log 2>../log/format.stderr.log"],
```
Comment (matthieu-foucault) [inline .watch-app-format.json]: `XO` prints a lot of spaces between the error message and the rule name, making the file hard to read with a limited width. This make it more readable. I can't find a way to set the eslint formatter in the `XO` config.
```suggestion
    "command": [
      "/usr/bin/env",
      "bash",
      "-c",
      "yarn run format | sed -E 's/\\s{2,}/\\t/g'  1>../log/format.stdout.log 2>../log/format.stderr.log"
    ],
```

File: .watch-status-schema.json [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+  "trigger",
+  "./",
+  {
+    "name": "schema",
+    "expression": [
+      "anyof",
+      ["match", "log/sqitch.status", "wholename"],
+      ["match", "app/package.json", "wholename"]
+    ],
+    "command": ["/usr/bin/env", "bash", "-c", "cd app && yarn run schema 1>../log/schema.stdout.log 2>../log/schema.stderr.log"],
```
Comment (matthieu-foucault) [inline .watch-status-schema.json]: `schema` was renamed to `build:schema`

```suggestion
    "command": [
      "/usr/bin/env",
      "bash",
      "-c",
      "cd app && yarn run build:schema 1>../log/schema.stdout.log 2>../log/schema.stderr.log"
    ],
```

File: documentation/developer-guide.md [hunk 6]
```diff
@@ -18,3 +18,22 @@ The `linter-xo-2` VSCode add XO support. The following VSCode settings will enab
...
+lifecycle jobs as needed. We define triggers using the [extended json syntax]
+and commit these configuration files to the project. To start all triggers, run
+`make watch` and to stop watching run `make unwatch`.
+
+The logs for all watched processes can be found in the `logs/` directory, and
+are overwritten on each restart of the triggered process. Certain additional
+state files are also stored in this folder. Notably, `sqitch.status` records
+the change hash currently deployed via watchman trigger to the local database
+and `server.pid` records the process id of the node app server run via watchman.
+
```
Comment (matthieu-foucault) [inline documentation/developer-guide.md]:
```suggestion

To easily monitor all of the log files, you can use a program such as [multitail]. For instance, if you want to output all of the `sdtdout` files in your terminal, split into two columns, you can do the following:
`$ multitail -f -s 2 *.stdout.log`

```

File: documentation/developer-guide.md [hunk 7]
```diff
@@ -18,3 +18,22 @@ The `linter-xo-2` VSCode add XO support. The following VSCode settings will enab
...
+Running `make unwatch` will stop both the node app server and the asdf-managed
+local postgres server.
+
+[Watchman]: https://facebook.github.io/watchman/
+[extended json syntax]: https://facebook.github.io/watchman/docs/cmd/trigger.html#extended-syntax
```
Comment (matthieu-foucault) [inline documentation/developer-guide.md]:
```suggestion
[extended json syntax]: https://facebook.github.io/watchman/docs/cmd/trigger.html#extended-syntax
[multitail]: https://linux.die.net/man/1/multitail
```

File: .watch-app-format.json (hunk 4 above)
Comment (wenzowski) [reply .watch-app-format.json]: the disadvantage to doing this is we break our row:column alignment

Comment (matthieu-foucault) [reply .watch-app-format.json]: I see. I guess we don't really need to watch this file anyway. Not until the linter passes at least.

**Pull request merged**